### DISCUSSION TITLE
Update BBQKees E32 V1.5 to 3.7.0

### DISCUSSION THREAD

**Discussion by crbble** [emsesp/EMS-ESP32#2186] [Show and tell]
Hi everyone,

I just wanted to share a short story about updating the "grey" BBQKees E32 1.5 (internal micro USB) to the newest version of EMS-ESP. The problem here is that coming from 3.4 the partition layout has changed and OTA is not possible anymore, previously OTA was fine.

The official variants to update are:

1. EMS-ESP-Flash-Tool: Huge binary file, I am not that paranoid but I was running out of options until I ran it. Then it segfaulted on an up to date Archlinux :-(
2. EMS-ESP Web installer: Only offers software for the E32 V2, which is 16 MB flash and therefore I do not expect it to work (and it actually didnt).
3. Web interface/OTA: Partition layout change.
4. EMS-ESP-cli installer: Only offers configs for E32 V2 16 MB as far as I saw and needs github command line tool and login.

5. Building from sourcecode: That worked. Took a while to setup platformio (and pull all dependencies), but _pio run -e s_4M --target upload --upload-port /dev/ttyUSB0_ did the job in the end. I flashed without external power supply, which worked fine enough.

Having flashed a lot of ESPs I was wondering if it is not possible to simply offer the flashfiles (partition, bootloader, flash etc.) and a one line esptool.py command line to write them to the correct address. Or did I miss this information somewhere?

Thanks a lot for this project and keep up the good work!

**Comment by proddy** [comment]
We used to do option 5, which included the `esptool.py` and the associated .bin files with instructions on how to upload using via the command line. I think we need to add that back. I don't know if @bbqkees has anything in his wiki on this?

@crbble fancy doing a PR?

**Comment by crbble** [comment]
Hi, thanks for the reply. 

Documentation from bbqkees starts here:
https://bbqkees-electronics.nl/wiki/gateway/firmware-update-and-downgrade.html#loading-firmware-via-the-esptool

And points here:
https://emsesp.org/Getting-Started/#uploading-the-firmware

This is a dead link (but I think since only a few days).

@proddy Can you outline the scope of the PR? Integrate the neccessary steps to create the files from the CI toolchain and place them somewhere downloadable and then add the esptool line to the docs?

**Reply by proddy** [reply]
I think what's needed is
- Adjust https://github.com/emsesp/EMS-ESP-Flasher-CLI and
  - add a new subfolder for the S32 and upload the binary (partition.bin, etc). I usually run `pio -v` and copy what pio uses with the `esptool` command. This will cover the 4MB variants. I should have written those scripts in Python but was lazy so used bash/sh. It's nasty, sorry.
  - include the EMS-ESP stable firmware bin files in the `firmware` folder, so users don't need to mess around with GitHub's gh CLI. Unless they want to use the `dev` branch.
- Make a comment on https://download.emsesp.org mentioning to use the CLI for 4MB versions.

This will work, but only covers Linux users. Windows users should still be able to use the Flash Tool.

You said the Flash Tool didn't work on ArchLinux - maybe you can find out why? I documented the build procedure in the readme

**Reply by crbble** [reply]
Hey, sorry for being slow, but I dug a little bit deeper into the EMS-ESP-Flasher topic.
Simply running the Ubuntu.tar GUI binary on recent Arch most likely has some library dependency problem:

> (No debugging symbols found in ./EMS-ESP-Flasher)
> (gdb) run
> Starting program: /home/rob/prog/python/EMS-ESP-Flasher/EMS-ESP-Flasher 
> [Thread debugging using libthread_db enabled]
> Using host libthread_db library "/usr/lib/libthread_db.so.1".
> [Detaching after fork from child process 64331]
> 
> Program received signal SIGSEGV, Segmentation fault.
> 0x00007ffff7df83f4 in ?? () from /usr/lib/libc.so.6
> (gdb) backtrace
> #0  0x00007ffff7df83f4 in ?? () from /usr/lib/libc.so.6
> #1  0x00007ffff7d9f120 in raise () from /usr/lib/libc.so.6
> #2  0x0000000000404224 in ?? ()
> #3  0x00007ffff7d87e08 in ?? () from /usr/lib/libc.so.6
> #4  0x00007ffff7d87ecc in __libc_start_main () from /usr/lib/libc.so.6
> #5  0x000000000040250f in ?? ()

Running the tool on command line at least gives me the help, due to the EMS-ESP already being at work I need to find some ESP32 to flash in my electronics box. Running the tool from source also works without problems, the dependencies are quite small, which is nice.

I would still be missing the correct flashfiles for bootloader and partitions with the tool, though.

Once I have some more time I will check where I can update the docs and add the 4MB versions to the downloader page.
